aptos-move/aptos-gas-calibration/src/math.rs

```diff
@@ -0,0 +1,350 @@
+            total_time = total_time + (a_ij * x_hat[(j, 0)]);
+            j = j + 1;
+        }
+        computed_running_time.push(total_time);
+        i = i + 1;
+        j = 0;
+    }
+
+    i = 0;
+
+    // compare w/ margin of error
+    let mut outliers: Vec<(usize, f64, f64, f64, bool)> = Vec::new();
+    let const_row = constant_matrix.nrows();
+    while i < const_row {
+        let a_ij = constant_matrix[(i, 0)];
+
+        let numerator = (a_ij - computed_running_time[i]).abs();
+        let denominator = computed_running_time[i];
+        if approx_eq!(f64, denominator, 0.0, ulps = 2) {
+            return Err(String::from("Division by zero"));
+        } else {
+            let diff = numerator.div(denominator);
+            if diff > MARGIN_OF_ERROR {
+                // append equation that is an outlier
+                outliers.push((i, a_ij, computed_running_time[i], diff, true));
+            } else {
+                // not outlier, but still display to the user
+                // the running time of that Calibration Function
+                outliers.push((i, a_ij, computed_running_time[i], diff, false));
+            }
+        }
+        i = i + 1;
+    }
+
+    Ok(outliers)
+}
+
+/// Find pivot columns if system of linear eq can't be solved
+///
+/// ### Arguments
+///
+/// * `matrix` - An input matrix to solve, typically the RREF'd matrix
+fn find_pivot_columns(matrix: &mut DMatrix<f64>) -> Vec<usize> {
+    let mut pivot_columns = Vec::new();
+    let ncols = matrix.ncols() - 1;
+
+    for j in 0..ncols {
+        let mut has_pivot = false;
+        for i in 0..matrix.nrows() {
+            if !approx_eq!(f64, matrix[(i, j)], 0.0, ulps = 2) {
+                has_pivot = true;
+                break;
+            }
+        }
+        if has_pivot {
+            pivot_columns.push(j);
+        }
+    }
+
+    pivot_columns
+}
+
+/// Reduced row echelon form (RREF)
+///
+/// ### Arguments
+///
+/// * `matrix` - A matrix to perform RREF
+fn rref(matrix: &mut DMatrix<f64>) {
+    let (nrows, ncols) = matrix.shape();
+    let mut lead = 0;
+
+    for r in 0..nrows {
+        if ncols <= lead {
+            break;
+        }
+
+        let mut i = r;
+
+        while approx_eq!(f64, matrix[(i, lead)], 0.0, ulps = 2) {
+            i += 1;
+
+            if nrows == i {
+                i = r;
+                lead += 1;
+
+                if ncols == lead {
+                    return;
+                }
+            }
+        }
+
+        if i != r {
+            matrix.swap_rows(i, r);
+        }
+
+        let pivot = matrix[(r, lead)];
+
+        for j in 0..ncols {
+            matrix[(r, j)] /= pivot;
+        }
+
+        for i in 0..nrows {
+            if i != r {
+                let factor = matrix[(i, lead)];
+                for j in 0..ncols {
+                    matrix[(i, j)] -= factor * matrix[(r, j)];
+                }
+            }
+        }
+
+        lead += 1;
+    }
+}
```

Comment:
TODO: partial pivoting for better numerical stability

Reply:
Addressed by updating RREF using partial pivoting.